helix-core/src/main/java/org/apache/helix/manager/zk/ZKHelixManager.java

```diff
@@ -1099,8 +1099,28 @@ public void handleStateChanged(KeeperState state) {
 
   @Override
   public void handleNewSession() throws Exception {
-    LOG.info(
-        "Handle new session, instance: " + _instanceName + ", type: " + _instanceType);
+    handleNewSession(null);
+  }
+
+  @Override
+  public void handleNewSession(final String sessionId) throws Exception {
+    /*
+     * When Zk session is re-established, some old events from the previous session might still be
+     * left in in the event queue. This may cause some issues related to session expiration.
+     * Theoretically, an old event coming from an old session will be re-triggered or reset in the
+     * new event, so we don't have to process the old event.
+     * Thus, if the session is expired, we will discard this session and will NOT run following
+     * tasks for this session. If session id is null, it means this operation is not session aware
+     * and so we don't have to check session expiration.
+     */
+    if (sessionId != null && !getSessionId().equals(sessionId)) {
```

Comment:
As we discussed offline, please move the logic change to your 2nd part PR. That makes this PR pure interface change.